src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/InterpolatedStringBuilder.cs

```diff
@@ -0,0 +1,596 @@
+        //   they'd all map to the object overloads as well.
+        // - Any reference type with an implicit cast to ROS<char> will fail to compile due to ambiguities between the overloads. string
+        //   is one such type, hence needing dedicated overloads for it that can be bound to more tightly.
+        //
+        // A middle ground we've settled on, which is likely to be the right approach for most other builders as well, would be the set:
+        //     (T, ...) with no constraint
+        //     (ReadOnlySpan<char>) and (ReadOnlySpan<char>, int)
+        //     (object, int alignment = 0, string? format = null)
+        //     (string) and (string, int)
+        // This would address most of the concerns, at the expense of:
+        // - Most reference types going through the generic code paths and so being a bit more expensive.
+        // - Nullable types being more expensive until https://github.com/dotnet/runtime/issues/50915 is addressed.
+        //   We could choose to add a T? where T : struct set of overloads if necessary.
+        // Strings don't require their own overloads here, but as they're expected to be very common and as we can
+        // optimize them in several ways (can copy the contents directly, don't need to do any interface checks, don't
+        // need to pay the shared generic overheads, etc.) we can add overloads specifically to optimize for them.
+        //
+        // Hole values are formatted according to the following policy:
+        // 1. If an IFormatProvider was supplied and it provides an ICustomFormatter, use ICustomFormatter.Format (even if the value is null).
+        // 2. If the type implements ISpanFormattable, use ISpanFormattable.TryFormat.
+        // 3. If the type implements IFormattable, use IFormattable.ToString.
+        // 4. Otherwise, use object.ToString.
+        // This matches the behavior of string.Format, StringBuilder.AppendFormat, etc.  The only overloads for which this doesn't
+        // apply is ReadOnlySpan<char>, which isn't supported by either string.Format nor StringBuilder.AppendFormat, but more
+        // importantly which can't be boxed to be passed to ICustomFormatter.Format.
+
+        #region TryFormatInterpolationHole T
+        /// <summary>Writes the specified value to the builder.</summary>
+        /// <param name="value">The value to write.</param>
+        public void TryFormatInterpolationHole<T>(T value)
+        {
+            // This method could delegate to TryFormatInterpolationHole with a null format, but explicitly passing
+            // default as the format to TryFormat helps to improve code quality in some cases when TryFormat is inlined,
+            // e.g. for Int32 it enables the JIT to eliminate code in the inlined method based on a length check on the format.
+
+            // If there's a custom formatter, always use it.
+            if (_customFormatter is not null)
```

Comment:
Random thought: If we expect that a call to the `InterpolatedStringBuilder` ctor and a call to this method will take place within the same frame, I wonder if some clever combination of inlining + JIT magic could allow the JIT to prove that this field is null for the common case, eliding this block entirely? I don't know how practical that optimization is.

Reply:
That'd be neat. :smile:  That said, I'd done some measurements on typical format strings to see how much these checks cost us (thinking maybe it'd be worth separating out the provider support into a separate builder), but it was small, maybe a couple percent if that.

As an aside, we already get some nice benefits from TryFormatBaseString being inlined, due to @egorbo's changes around propagating data from const strings... TryFormatBaseString is almost always used with const strings, because that's what the compiler targets for the const portions of the format strings.